apps/agent/contracts/Agent.sol

```diff
@@ -0,0 +1,157 @@
+    * @param _target Address where the action is being executed
+    * @param _ethValue Amount of ETH from the contract that is sent with the action
+    * @param _data Calldata for the action
+    * @return Exits call frame forwarding the return data of the executed call (either error or success data)
+    */
+    function execute(address _target, uint256 _ethValue, bytes _data)
+        external // This function MUST always be external as the function performs a low level return, exiting the Agent app execution context
+        authP(EXECUTE_ROLE, arr(_target, _ethValue, uint256(getSig(_data)))) // bytes4 casted as uint256 sets the bytes as the LSBs
+    {
+        bool result = _target.call.value(_ethValue)(_data);
+
+        if (result) {
+            emit Execute(msg.sender, _target, _ethValue, _data);
+        }
+
+        assembly {
+            let size := returndatasize
+            let ptr := mload(0x40)
+            returndatacopy(ptr, 0, size)
+
+            // revert instead of invalid() bc if the underlying call failed with invalid() it already wasted gas.
+            // if the call returned error data, forward it
+            switch result case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+
+    /**
+    * @notice Set `_designatedSigner` as the designated signer of the app, which will be able to sign messages on behalf of the app
+    * @param _designatedSigner Address that will be able to sign messages on behalf of the app
+    */
+    function setDesignatedSigner(address _designatedSigner)
+        external
+        authP(DESIGNATE_SIGNER_ROLE, arr(_designatedSigner))
+    {
+        // Prevent an infinite loop by setting the app itself as its designated signer.
+        // An undetectable loop can be created by setting a different contract as the
+        // designated signer which calls back into `isValidSignature`.
+        // Given that `isValidSignature` is always called with just 50k gas, the max
+        // damage of the loop is wasting 50k gas.
+        require(_designatedSigner != address(this), ERROR_DESIGNATED_TO_SELF);
+
+        address oldDesignatedSigner = designatedSigner;
+        designatedSigner = _designatedSigner;
+
+        emit SetDesignatedSigner(msg.sender, oldDesignatedSigner, _designatedSigner);
+    }
+
+    /**
+    * @notice Pre-sign hash `_hash`
+    * @param _hash Hash that will be considered signed regardless of the signature checked with 'isValidSignature()'
+    */
+    function presignHash(bytes32 _hash)
+        external
+        authP(ADD_PRESIGNED_HASH_ROLE, arr(_hash))
+    {
+        isPresigned[_hash] = true;
+
+        emit PresignHash(msg.sender, _hash);
+    }
+
+    function isForwarder() external pure returns (bool) {
+        return true;
+    }
+
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool) {
+        return
+            interfaceId == ERC1271_INTERFACE_ID ||
```

Comment:
ERC1271 is still a draft. Using unfinished ERC may lead to making multiple updates to the system before it stabilizes. And making serious upgrades is always a security consideration.